espnet/nets/chain/rnn.py

```diff
@@ -0,0 +1,143 @@
+
+
+# TODO(watanabe) explanation of BLSTMP
+class BLSTMP(chainer.Chain):
+    def __init__(self, idim, elayers, cdim, hdim, subsample, dropout):
+        super(BLSTMP, self).__init__()
+        with self.init_scope():
+            for i in six.moves.range(elayers):
+                if i == 0:
+                    inputdim = idim
+                else:
+                    inputdim = hdim
+                setattr(self, "bilstm%d" % i, L.NStepBiLSTM(
+                    1, inputdim, cdim, dropout))
+                # bottleneck layer to merge
+                setattr(self, "bt%d" % i, L.Linear(2 * cdim, hdim))
+
+        self.elayers = elayers
+        self.cdim = cdim
+        self.subsample = subsample
+
+    def __call__(self, xs, ilens):
+        """BLSTMP forward
+
+        :param xs:
+        :param ilens:
+        :return:
+        """
+        logging.info(self.__class__.__name__ + ' input lengths: ' + str(ilens))
+
+        for layer in six.moves.range(self.elayers):
+            hy, cy, ys = self['bilstm' + str(layer)](None, None, xs)
+            # ys: utt list of frame x cdim x 2 (2: means bidirectional)
+            # TODO(watanabe) replace subsample and FC layer with CNN
+            ys, ilens = _subsamplex(ys, self.subsample[layer + 1])
+            # (sum _utt frame_utt) x dim
+            ys = self['bt' + str(layer)](F.vstack(ys))
+            xs = F.split_axis(ys, np.cumsum(ilens[:-1]), axis=0)
+            del hy, cy
+
+        # final tanh operation
+        xs = F.split_axis(F.tanh(F.vstack(xs)), np.cumsum(ilens[:-1]), axis=0)
+
+        # 1 utterance case, it becomes an array, so need to make a utt tuple
+        if not isinstance(xs, tuple):
+            xs = [xs]
+
+        return xs, ilens  # x: utt list of frame x dim
+
+
+class BLSTM(chainer.Chain):
+    def __init__(self, idim, elayers, cdim, hdim, dropout):
+        super(BLSTM, self).__init__()
+        with self.init_scope():
+            self.nblstm = L.NStepBiLSTM(elayers, idim, cdim, dropout)
+            self.l_last = L.Linear(cdim * 2, hdim)
+
+    def __call__(self, xs, ilens):
+        """BLSTM forward
+
+        :param xs:
+        :param ilens:
+        :return:
+        """
+        logging.info(self.__class__.__name__ + ' input lengths: ' + str(ilens))
+        # need to move ilens to cpu
+        ilens = cuda.to_cpu(ilens)
+        hy, cy, ys = self.nblstm(None, None, xs)
+        ys = self.l_last(F.vstack(ys))  # (sum _utt frame_utt) x dim
+        xs = F.split_axis(ys, np.cumsum(ilens[:-1]), axis=0)
+        del hy, cy
+
+        # final tanh operation
+        xs = F.split_axis(F.tanh(F.vstack(xs)), np.cumsum(ilens[:-1]), axis=0)
+
+        # 1 utterance case, it becomes an array, so need to make a utt tuple
+        if not isinstance(xs, tuple):
+            xs = [xs]
+
+        return xs, ilens  # x: utt list of frame x dim
+
+
+# TODO(watanabe) explanation of VGG2L, VGG2B (Block) might be better
+class VGG2L(chainer.Chain):
```

Comment:
vgg2l is not rnn and naming this module as  rnn.py is misleading. I think we can name it encoders.py